irc/accounts.go

```diff
@@ -30,14 +31,24 @@ const (
 	keyAccountRegTime          = "account.registered.time %s"
 	keyAccountCredentials      = "account.credentials %s"
 	keyAccountAdditionalNicks  = "account.additionalnicks %s"
+	keyAccountVHost            = "account.vhost %s"
 	keyCertToAccount           = "account.creds.certfp %s"
+
+	keyVHostQueueAcctToId = "vhostQueue %s"
+	vhostRequestIdx       = "vhostQueue"
 )
 
 // everything about accounts is persistent; therefore, the database is the authoritative
 // source of truth for all account information. anything on the heap is just a cache
 type AccountManager struct {
+	// XXX these are up here so they can be aligned to a 64-bit boundary, please forgive me
+	// autoincrementing ID for vhost requests:
+	vhostRequestID           uint64
```

Comment:
For vhost requests, do we need to have this sort of ID or can we simply use the account name as the request ID? (and/or just put the information directly in `keyAccountVHost`). Each account can only have a single request ongoing at a time, seems a bit weird to link it in this sorta way.

Reply:
The main goal of this is for `WAITING` to be able to list the requests in the order they were created. I used an autoincrementing ID because I'm stuck in a SQL mindset :-) The outcome I wanted to avoid is having `WAITING` always display in alphabetical order.

The buntdb way would be to use a JSON index on the `LastRequestTime` field of the vhost blob (possibly after massaging it into a form where it can be compared with string comparisons). I'm not super comfortable with the JSON index feature though.

Reply:
Just to clarify the current implementation:

1. Each new request is assigned an autoincrementing ID (the current position of the ID counter is maintained on the heap, but persisted to the DB)
1. Requests are written as buntdb key-values, mapping the account name (plus a prefix, so it ends up being something like `vhostQueue myAccountName`) to the autoincrementing ID as a string (e.g., `53`).
1. A buntdb numeric-order index is created on these key-values: `tx.CreateIndex(vhostRequestIdx, fmt.Sprintf(keyVHostQueueAcctToId, "*"), buntdb.IndexInt)`
1. `HS WAITING` looks up the first 10 requests in the index, then does a separate lookup for the actual requested vhost and other data for each of them.

The main scalability concern here is that the buntdb index has to be built from scratch every time the server starts up, and this is O(n) in the number of pending requests. Apart from that, I think the current implementation has most of the characteristics we might want. (I might add some defensive code so that if there's a bug where the queue gets out of sync with the source-of-truth JSON blobs, `HS REJECT` will unconditionally delete the queue entry for the target account.)